File: scripts/prepare-transparent-catalog-media.mjs
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { basename, resolve } from "node:path";
import sharp from "sharp";

const report = JSON.parse(await readFile(resolve(".tmp/catalog-media-audit.json"), "utf8"));
const outputDir = resolve(".tmp/catalog-transparent");
await mkdir(outputDir, { recursive: true });

function canonicalAssets(product) {
  const preferred = new Map(product.variants.filter((variant) => variant.image).map((variant) => [variant.id, variant.image]));
  const grouped = new Map();
  const out = product.assets.filter((asset) => !asset.variantId);
  product.assets.filter((asset) => asset.variantId).forEach((asset) => {
    if (!grouped.has(asset.variantId)) grouped.set(asset.variantId, []);
    grouped.get(asset.variantId).push(asset);
  });
  grouped.forEach((assets, variantId) => out.push(assets.find((asset) => asset.path === preferred.get(variantId)) || assets[0]));
  return out;
}

async function removeConnectedNeutralBackground(input) {
  const { data, info } = await sharp(input).rotate().ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  const { width, height } = info;
  const pixels = width * height;
  const seen = new Uint8Array(pixels);
  const queue = new Int32Array(pixels);
  let head = 0;
  let tail = 0;
  const isBackground = (pixel) => {
    const offset = pixel * 4;
    const r = data[offset], g = data[offset + 1], b = data[offset + 2], a = data[offset + 3];
    const high = Math.max(r, g, b), low = Math.min(r, g, b), avg = (r + g + b) / 3;
    return a < 20 || (high - low <= 24 && avg >= 216);
  };
  const enqueue = (pixel) => {
    if (pixel < 0 || pixel >= pixels || seen[pixel] || !isBackground(pixel)) return;
    seen[pixel] = 1;
    queue[tail++] = pixel;
  };
  for (let x = 0; x < width; x += 1) { enqueue(x); enqueue((height - 1) * width + x); }
  for (let y = 1; y < height - 1; y += 1) { enqueue(y * width); enqueue(y * width + width - 1); }
  while (head < tail) {
    const pixel = queue[head++], x = pixel % width;
    if (x) enqueue(pixel - 1);
    if (x + 1 < width) enqueue(pixel + 1);
    if (pixel >= width) enqueue(pixel - width);
    if (pixel + width < pixels) enqueue(pixel + width);
  }
  for (let pixel = 0; pixel < pixels; pixel += 1) {
    const offset = pixel * 4;
    if (seen[pixel]) data[offset + 3] = 0;
    if (data[offset + 3] === 0) data[offset] = data[offset + 1] = data[offset + 2] = 0;
  }
  return sharp(data, { raw: info }).png().toBuffer();
}

const outputs = [];
for (const product of report.products) {
  for (const asset of canonicalAssets(product)) {
    const affected = asset.analysis.whiteBackdrop || asset.analysis.brightNeutralBackdrop;
    const isPrimary = asset.role === "icon";
    const isVariant = Boolean(asset.variantId && product.variants.some((variant) => variant.id === asset.variantId && variant.image === asset.path));
    if (!affected || (!isPrimary && !isVariant)) continue;
    const transparent = await removeConnectedNeutralBackground(asset.file);
    const subject = await sharp(transparent).trim({ background: { r: 0, g: 0, b: 0, alpha: 0 } }).resize({ width: 720, height: 720, fit: "inside", withoutEnlargement: false }).png().toBuffer();
    const filename = `${product.id}-${isPrimary ? "icon" : `variant-${asset.variantId}`}.png`;
    const output = resolve(outputDir, filename);
    await sharp({ create: { width: 1200, height: 1200, channels: 4, background: { r: 0, g: 0, b: 0, alpha: 0 } } })
      .composite([{ input: subject, gravity: "center" }])
      .png({ compressionLevel: 9, adaptiveFiltering: true })
      .toFile(output);
    const stats = await sharp(output).stats();
    outputs.push({ productId: product.id, title: product.title, role: asset.role, variantId: asset.variantId, source: asset.path, output, filename: basename(output), alphaMin: stats.channels[3]?.min, alphaMax: stats.channels[3]?.max });
  }
}

const manifest = resolve(".tmp/catalog-transparent-manifest.json");
await writeFile(manifest, JSON.stringify({ generatedAt: new Date().toISOString(), outputs }, null, 2));
console.log(JSON.stringify({ manifest, images: outputs.length, icons: outputs.filter((item) => item.variantId == null).length, variants: outputs.filter((item) => item.variantId != null).length }, null, 2));

File: scripts/prepare-transparent-catalog-media.py
import json
from pathlib import Path

from PIL import Image
from rembg import new_session, remove

root = Path(__file__).resolve().parent.parent
audit = json.loads((root / ".tmp" / "catalog-media-audit.json").read_text(encoding="utf-8"))
manifest_path = root / ".tmp" / "catalog-transparent-manifest.json"
manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
files = {
    (product["id"], asset["path"]): Path(asset["file"])
    for product in audit["products"]
    for asset in product["assets"]
}

session = new_session()
for index, item in enumerate(manifest["outputs"], start=1):
    source = files[(item["productId"], item["source"])]
    image = Image.open(source).convert("RGBA")
    cutout = remove(image, session=session, alpha_matting=False)
    alpha = cutout.getchannel("A")
    bbox = alpha.getbbox()
    if not bbox:
        raise RuntimeError(f"Background removal produced an empty image for {source}")
    subject = cutout.crop(bbox)
    subject.thumbnail((720, 720), Image.Resampling.LANCZOS)
    canvas = Image.new("RGBA", (1200, 1200), (0, 0, 0, 0))
    canvas.alpha_composite(subject, ((1200 - subject.width) // 2, (1200 - subject.height) // 2))
    output = Path(item["output"])
    output.parent.mkdir(parents=True, exist_ok=True)
    canvas.save(output, "PNG", optimize=True, compress_level=9)
    if index % 10 == 0:
        print(f"processed {index}/{len(manifest['outputs'])}", flush=True)

print(json.dumps({"processed": len(manifest["outputs"]), "outputDir": str(Path(manifest["outputs"][0]["output"]).parent)}))
